ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs

```diff
@@ -320,6 +302,32 @@ instance Crypto c => ShowQuery (Query (ShelleyBlock c)) where
   showResult GetProposedPParamsUpdates      = show
   showResult GetStakeDistribution           = show
 
+{-------------------------------------------------------------------------------
+  ValidateEnvelope
+-------------------------------------------------------------------------------}
+
+instance Crypto c => ValidateEnvelope (ShelleyBlock c) where
+  type OtherHeaderEnvelopeError (ShelleyBlock c) =
+    ShelleyOtherHeaderEnvelopeError c
+
+  validateEnvelope cfg ledgerView oldTip hdr = do
+      -- In addition to the default 'validateEnvelope' ...
+      defaultValidateEnvelope oldTip hdr
+      -- ... perform the @chainChecks@ that are part of the @CHAIN@ rule.
+      withExcept (OtherHeaderEnvelopeError . ShelleyOtherHeaderEnvelopeError) $
+        SL.chainChecks globals pparams (shelleyHeaderRaw hdr)
+    where
+      ShelleyLedgerConfig globals = configLedger cfg
+      pparams = SL.lvProtParams ledgerView
+
+newtype ShelleyOtherHeaderEnvelopeError c = ShelleyOtherHeaderEnvelopeError {
+      unShelleyOtherHeaderEnvelopeError :: STS.PredicateFailure (STS.CHAIN c)
+    }
+  deriving (Eq, Show)
+  -- TODO fix thunks
+  deriving NoUnexpectedThunks
+    via OnlyCheckIsWHNF "ShelleyOtherHeaderEnvelopeError" (ShelleyOtherHeaderEnvelopeError c)
```

Comment:
Let's make this an orphan instance of `NoUnexpectedThunks` instead, so that I'm forced to remove it when I add the right one in `cardano-ledger-specs`.